frame/support/src/traits/hooks.rs

```diff
@@ -447,60 +479,43 @@ mod tests {
 			}
 			fn post_upgrade(state: Vec<u8>) -> Result<(), &'static str> {
 				let s: bool = Decode::decode(&mut state.as_slice()).unwrap();
+				Test3Assertions::mutate(|val| *val += 1);
 				assert_eq!(s, true);
 				Ok(())
 			}
 		}
 
-		type TestEmpty = ();
-		let origin_state = <TestEmpty as OnRuntimeUpgrade>::pre_upgrade().unwrap();
-		let states: Vec<Vec<u8>> = Decode::decode(&mut origin_state.as_slice()).unwrap();
-		assert!(states.is_empty());
-		<TestEmpty as OnRuntimeUpgrade>::post_upgrade(origin_state).unwrap();
-
-		type Test1Tuple = (Test1,);
-		let origin_state = <Test1Tuple as OnRuntimeUpgrade>::pre_upgrade().unwrap();
-		let states: Vec<Vec<u8>> = Decode::decode(&mut origin_state.as_slice()).unwrap();
-		assert_eq!(states.len(), 1);
-		assert_eq!(
-			<String as Decode>::decode(&mut states[0].as_slice()).unwrap(),
-			"Test1".to_owned()
-		);
-		<Test1Tuple as OnRuntimeUpgrade>::post_upgrade(origin_state).unwrap();
-
-		type Test123 = (Test1, Test2, Test3);
-		let origin_state = <Test123 as OnRuntimeUpgrade>::pre_upgrade().unwrap();
-		let states: Vec<Vec<u8>> = Decode::decode(&mut origin_state.as_slice()).unwrap();
-		assert_eq!(
-			<String as Decode>::decode(&mut states[0].as_slice()).unwrap(),
-			"Test1".to_owned()
-		);
-		assert_eq!(<u32 as Decode>::decode(&mut states[1].as_slice()).unwrap(), 100u32);
-		assert_eq!(<bool as Decode>::decode(&mut states[2].as_slice()).unwrap(), true);
-		<Test123 as OnRuntimeUpgrade>::post_upgrade(origin_state).unwrap();
-
-		type Test321 = (Test3, Test2, Test1);
-		let origin_state = <Test321 as OnRuntimeUpgrade>::pre_upgrade().unwrap();
-		let states: Vec<Vec<u8>> = Decode::decode(&mut origin_state.as_slice()).unwrap();
-		assert_eq!(<bool as Decode>::decode(&mut states[0].as_slice()).unwrap(), true);
-		assert_eq!(<u32 as Decode>::decode(&mut states[1].as_slice()).unwrap(), 100u32);
-		assert_eq!(
-			<String as Decode>::decode(&mut states[2].as_slice()).unwrap(),
-			"Test1".to_owned()
-		);
-		<Test321 as OnRuntimeUpgrade>::post_upgrade(origin_state).unwrap();
-
-		type TestNested123 = (Test1, (Test2, Test3));
-		let origin_state = <TestNested123 as OnRuntimeUpgrade>::pre_upgrade().unwrap();
-		let states: Vec<Vec<u8>> = Decode::decode(&mut origin_state.as_slice()).unwrap();
-		assert_eq!(
-			<String as Decode>::decode(&mut states[0].as_slice()).unwrap(),
-			"Test1".to_owned()
-		);
-		// nested state for (Test2, Test3)
-		let nested_states: Vec<Vec<u8>> = Decode::decode(&mut states[1].as_slice()).unwrap();
-		assert_eq!(<u32 as Decode>::decode(&mut nested_states[0].as_slice()).unwrap(), 100u32);
-		assert_eq!(<bool as Decode>::decode(&mut nested_states[1].as_slice()).unwrap(), true);
-		<TestNested123 as OnRuntimeUpgrade>::post_upgrade(origin_state).unwrap();
+		TestExternalities::default().execute_with(|| {
```

Comment:
This test is not quite needed anymore as we don't combine the outputs anymore.

Reply:
I mean, it is nice that we're testing `impl_trait_for_tuples`, but I think it should be more meaningful.

For example, with the new logic, I would test the following: 

1. if `feature = try-runtime`, pre-migrate has been run. You can assert this by writing to some static value in `pre_migrate` 
2. if if `not(feature = try-runtime)`, they don't happen. You can test this by making sure a `OnRuntimeUpgrade` that panics `pre_upgrade` runs fine in `not(feature = try-runtime)`.

Reply:
I don’t see a reason to introduce those tbh. try-runtime methods are feature-gated, so are the tests. We can’t test feature-gated stuff in a non-feature-gated test. 
We have assertions in the test implementations of OnRuntimeUpgrade trait and I think it’s totally valid to test all those scenarios to see that things have been executed with nested tuples, for example.